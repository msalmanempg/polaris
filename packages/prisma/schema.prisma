//******** AUTO GENERATED FILE, DO NOT EDIT.  *********

datasource db {
    provider = "postgresql"
    url      = env("DATABASE_URL")
}

generator client {
    provider = "prisma-client-js"
}

generator dbml {
    provider = "prisma-dbml-generator"
}


model Agreement {
  id                   Int                 @id @default(autoincrement())
  leadId               Int                 @map("lead_id")
  agreementNumber      String              @unique @map("agreement_number")
  publishedPrice       Int                 @map("published_price")
  bookingPrice         Int                 @map("booking_price")
  discountValue        Int?                @map("discount_value")
  posessionValue       Int                 @map("posession_value")
  payments             ExpectedPayment[]
  customers            AgreementCustomer[]
  salesPerson          SalePerson[]
  nominees             Nominee[]
  Unit                 Unit                @relation(fields: [unitId], references: [id])
  unitId               Int                 @map("unit_id")
  status               agreementStatus
  bookingDate          DateTime            @map("booking_date")
  notarisedDate        DateTime            @map("notarised_date")
  agreementDate        DateTime            @map("agreement_date")
  installmentStartDate DateTime            @map("installment_start_date")
  installmentEndDate   DateTime            @map("installment_end_date")
  createdAt            DateTime            @default(now()) @map("created_at")
  updatedAt            DateTime            @updatedAt @map("updated_at")

  @@map("agreements")
}

model DraftAgreement {
  id              Int      @id @default(autoincrement())
  Unit            Unit     @relation(fields: [unitId], references: [id])
  unitId          Int      @map("unit_id")
  agreementNumber String   @unique @map("agreement_number")
  data            Json     @map("data")
  createdAt       DateTime @default(now()) @map("created_at")
  updatedAt       DateTime @updatedAt @map("updated_at")

  @@map("draft_agreements")
}

model SalePerson {
  id          Int       @id @default(autoincrement())
  designation String
  fullName    String    @map("full_name")
  email       String
  govtId      String    @map("govt_id")
  phoneNumber String    @map("phone_number")
  sortLevel   Int       @map("sort_level")
  Agreement   Agreement @relation(fields: [agreementId], references: [id])
  agreementId Int       @map("agreement_id")
  createdAt   DateTime  @default(now()) @map("created_at")
  updatedAt   DateTime  @updatedAt @map("updated_at")

  @@map("sale_persons")
}

model AgreementCustomer {
  id                   Int                @id @default(autoincrement())
  type                 customerType       @map("customer_type")
  Customer             Customer           @relation(fields: [customerId], references: [id])
  customerId           Int                @map("customer_id")
  Agreement            Agreement          @relation(fields: [agreementId], references: [id])
  agreementId          Int                @map("agreement_id")
  isPrimary            Boolean            @default(true)
  meta                 Json?
  email                String
  primaryPhoneNumber   String             @map("primary_phone_number")
  secondaryPhoneNumber String?            @map("secondary_phone_number")
  address              String
  Companies            AgreementCompany[]
  createdAt            DateTime           @default(now()) @map("created_at")
  updatedAt            DateTime           @updatedAt @map("updated_at")

  @@map("agreement_customers")
}

model AgreementCompany {
  id                  Int               @id @default(autoincrement())
  AgreementCustomer   AgreementCustomer @relation(fields: [agreementCustomerId], references: [id])
  agreementCustomerId Int               @map("agreement_customer_id")
  name                String
  ntn                 String
  registrationNumber  String            @unique @map("registration_number")
  address             String
  city                String
  email               String
  contactNumber       String            @map("contact_number")
  meta                Json?
  createdAt           DateTime          @default(now()) @map("created_at")
  updatedAt           DateTime          @updatedAt @map("updated_at")

  @@map("agreement_companies")
}

model Nominee {
  id            Int        @id @default(autoincrement())
  govtId        String     @map("govt_id")
  govtIdType    govtIdType @map("govt_id_card")
  nomineeFor    String     @map("nominee_for")
  fullName      String     @map("full_name")
  contactNumber String     @map("contact_number")
  email         String
  relationship  String
  Agreement     Agreement  @relation(fields: [agreementId], references: [id])
  agreementId   Int        @map("agreement_id")
  meta          Json?
  createdAt     DateTime   @default(now()) @map("created_at")
  updatedAt     DateTime   @updatedAt @map("updated_at")

  @@map("nominees")
}

model AgreementAttachment {
  id           Int                   @id @default(autoincrement())
  agreementId  Int                   @map("agreement_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         agreementDocumentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("agreement_attachments")
}

model NomineeAttachment {
  id           Int                   @id @default(autoincrement())
  nomineeId    Int                   @map("nominee_id")
  Attachment   Attachment            @relation(fields: [attachmentId], references: [id])
  attachmentId Int
  type         nomineeAttachmentType
  createdAt    DateTime              @default(now()) @map("created_at")
  updatedAt    DateTime              @updatedAt @map("updated_at")

  @@map("nominee_attachment")
}

enum govtIdType {
  cnic
  nicop
  poc
}

enum customerType {
  individual
  company
}

enum agreementStatus {
  draft
  review
  active
  transferred
  terminated
}

enum agreementDocumentType {
  agreement
  booking_form
  payment_plan
  floor_plan
  other
}

enum nomineeAttachmentType {
  id_front
  id_back
  profile_picture
  other
}


model Attachment {
  id                   Int                   @id @default(autoincrement())
  name                 String
  key                  String
  customerAttachments  CustomerAttachment[]
  nomineeAttachments   NomineeAttachment[]
  agreementAttachments AgreementAttachment[]
  createdAt            DateTime              @default(now()) @map("created_at")
  updatedAt            DateTime              @updatedAt @map("updated_at")

  @@map("attachments")
}


model Customer {
    id             Int                  @id @default(autoincrement())
    fullName       String               @map("full_name")
    govtId         String               @map("govt_id")
    govtIdType     govtIdType           @map("govt_id_type")
    relationType   relationType         @map("relation_type")
    guardianName   String               @map("guardian_name")
    nationality    String
    passportNumber String?              @map("passport_number")
    dateOfBirth    DateTime             @map("date_of_birth")
    gender         Gender
    meta           Json?
    attachments    CustomerAttachment[]
    agreements     AgreementCustomer[]
    createdAt      DateTime             @default(now()) @map("created_at")
    updatedAt      DateTime             @updatedAt @map("updated_at")

    @@unique([govtId, govtIdType], name: "Customer_govtId_govtIdType_unique_constraint")
    @@map(name: "customers")
}

model CustomerAttachment {
    id           Int        @id @default(autoincrement())
    Attachment   Attachment @relation(fields: [attachmentId], references: [id])
    attachmentId Int
    Customer     Customer   @relation(fields: [customerId], references: [id])
    customerId   Int
    createdAt    DateTime   @default(now()) @map("created_at")
    updatedAt    DateTime   @updatedAt @map("updated_at")

    @@map("customer_attachments")
}

enum relationType {
    father
    husband
}

enum Gender {
    male
    female
}

enum customerDocumentType {
    id_front
    id_back
    profile_picture
    other
}


model ExpectedPayment {
  id          Int                   @id @default(autoincrement())
  amount      Int
  percentage  Int?
  paymentDate DateTime              @map("payment_date")
  type        expectedPaymentType
  Agreement   Agreement             @relation(fields: [agreementId], references: [id])
  agreementId Int                   @map("agreement_id")
  meta        Json?
  dueDate     DateTime              @map("due_date")
  status      expectedPaymentStatus
  createdAt   DateTime              @default(now()) @map("created_at")
  updatedAt   DateTime              @updatedAt @map("updated_at")

  @@map("expected_payments")
}

enum expectedPaymentType {
  down_payment
  installment
  advance
  posession
}

enum expectedPaymentStatus {
  paid
  pending
}


model Project {
  id             Int                 @id @default(autoincrement())
  name           String
  address        String
  city           String
  province       String
  country        String
  longitude      Float
  latitude       Float
  email          String
  phone          String
  url            String
  completionDate DateTime?           @map("completion_date")
  logo           String
  meta           Json?
  bankDetails    ProjectBankDetail[]
  units          Unit[]
  createdAt      DateTime            @default(now()) @map("created_at")
  updatedAt      DateTime            @updatedAt @map("updated_at")

  @@map("projects")
}

model ProjectBankDetail {
  id            Int      @id @default(autoincrement())
  bankName      String   @map("bank_name")
  accountTitle  String   @map("account_title")
  accountNumber String   @map("account_number")
  iban          String
  swiftCode     String   @map("swift_code")
  branchAddress String   @map("branch_address")
  branchCode    String   @map("branch_code")
  meta          Json?
  Project       Project? @relation(fields: [projectId], references: [id])
  projectId     Int?
  createdAt     DateTime @default(now()) @map("created_at")
  updatedAt     DateTime @updatedAt @map("updated_at")

  @@map("project_bank_details")
}


model Unit {
  id             Int              @id @default(autoincrement())
  unitNumber     String           @unique @map("unit_number")
  location       String
  type           unitType
  bed            Int
  basePrice      Int              @map("base_price")
  publishedPrice Int              @map("published_price")
  netArea        Int              @map("net_area")
  grossArea      Int              @map("gross_area")
  meta           Json?
  status         unitStatus
  Project        Project          @relation(fields: [projectId], references: [id])
  projectId      Int              @map("project_id")
  agreements     Agreement[]
  drafts         DraftAgreement[]
  completionDate DateTime?        @map("completion_date")
  createdAt      DateTime         @default(now()) @map("created_at")
  updatedAt      DateTime         @updatedAt @map("updated_at")

  @@map(name: "units")
}

enum unitType {
  residential
  commercial
}

enum unitStatus {
  booked
  sold
  available
}
