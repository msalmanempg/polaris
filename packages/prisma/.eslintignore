**/__snapshots__/
**/.ng_build/
**/build/
**/coverage/
**/dist/
**/lib/
**/node_modules/
jest.config.js
*.d.ts