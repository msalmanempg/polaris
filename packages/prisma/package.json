{
  "name": "@polaris/prisma",
  "version": "1.0.0",
  "description": "data layer for Polaris",
  "main": "lib/index.js",
  "repository": "https://github.com/SectorLabs/polaris/tree/main/packages/prisma",
  "license": "MIT",
  "scripts": {
    "build": "yarn ttsc -p tsconfig.json",
    "lint:errors": "eslint --quiet --ext .js,.ts ./",
    "lint:types": "ttsc --noEmit"
  }
}
