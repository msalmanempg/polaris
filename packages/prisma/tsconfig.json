{
  "compilerOptions": {
    "allowJs": true,
    "module": "commonjs",
    "esModuleInterop": true,
    "moduleResolution": "node",
    "skipLibCheck": true,
    "isolatedModules": true,
    "forceConsistentCasingInFileNames": true,
    "lib": ["es2017", "dom"],
    "target": "es2017",
    "types": ["node"],
    "strict": true,
    "baseUrl": ".",
    "paths": {
      "prisma/*": ["."]
    },
    "outDir": "./lib/",
    "sourceMap": true,
    "inlineSources": true,
    "sourceRoot": "/",
    "plugins": [{ "transform": "@zerollup/ts-transform-paths" }]
  },
}
