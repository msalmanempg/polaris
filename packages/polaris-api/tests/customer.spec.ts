import { Customer, prisma } from "@polaris/prisma";
import { customers } from "./support/fixtures/customers";
import { testExpressApp } from "./support/requestFactory";

describe("customers", () => {
  let createdCustomers: Customer[];

  beforeEach(async () => {
    createdCustomers = await Promise.all(
      customers.map((customer) => {
        return prisma.customer.create({
          data: customer as any,
        });
      })
    );
  });

  afterEach(async () => {
    await prisma.agreementCustomer.deleteMany({});
    await prisma.customer.deleteMany({
      where: {
        id: {
          in: createdCustomers.map((customer) => customer.id),
        },
      },
    });
  });

  it("Should create a new Customer", async () => {
    const customer = {
      fullName: "Test Customer",
      govtId: "88888-8888888-8",
      govtIdType: "cnic",
      relationType: "father",
      guardianName: "Test Customer Guardian",
      nationality: "Pakistani",
      passportNumber: "11111-1111111-1",
      dateOfBirth: "2021-02-19T05:20:01.269Z",
      gender: "male",
      meta: {},
    };

    const { body } = await testExpressApp().post("/api/customer/add").send(customer).expect(201);

    expect(body).toMatchSnapshot({
      id: expect.any(Number),
      createdAt: expect.any(String),
      updatedAt: expect.any(String),
    });
  });

  it("Should List all Customers", async () => {
    const params = {
      pageOffset: 0,
      pageSize: 10,
    };

    const { body } = await testExpressApp().get("/api/customer/").query(params).expect(200);

    const createdCustomersIds = new Set(createdCustomers.map((customer) => customer.id));
    const customers = body
      .filter((customer: any) => createdCustomersIds.has(customer.id))
      .sort((a: { govtId: string }, b: { govtId: string }) => a.govtId.localeCompare(b.govtId));

    customers.map((customer: any) => {
      expect(customer).toMatchSnapshot({
        id: expect.any(Number),
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
      });
    });
  });

  it("Should update existing Customer", async () => {
    const existingCustomer = createdCustomers[0];

    const customer = {
      fullName: "Test Customer",
      govtId: "99999-9999999-9",
      govtIdType: "cnic",
      relationType: "father",
      guardianName: "Test Customer Guardian",
      nationality: "Pakistani",
      passportNumber: "11111-1111111-1",
      dateOfBirth: "2021-02-19T05:20:01.269Z",
      gender: "male",
      meta: {},
    };

    await testExpressApp()
      .put(`/api/customer/update/${existingCustomer.id}`)
      .send(customer)
      .expect(204);
  });

  it("Should delete existing Customer", async () => {
    const existingCustomer = createdCustomers[0];

    await testExpressApp().delete(`/api/customer/delete/${existingCustomer.id}`).expect(204);
  });
});
