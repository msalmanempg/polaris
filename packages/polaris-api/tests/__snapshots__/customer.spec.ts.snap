// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customers Should List all Customers 1`] = `
Object {
  "createdAt": Any<String>,
  "dateOfBirth": "2021-02-19T05:20:01.269Z",
  "fullName": "Test Customer",
  "gender": "male",
  "govtId": "12345-1234567-1",
  "govtIdType": "cnic",
  "guardianName": "Test Customer Guardian",
  "id": Any<Number>,
  "meta": Object {},
  "nationality": "Pakistani",
  "passportNumber": "11111-1111111-1",
  "relationType": "father",
  "updatedAt": Any<String>,
}
`;

exports[`customers Should List all Customers 2`] = `
Object {
  "createdAt": Any<String>,
  "dateOfBirth": "2021-02-19T05:20:01.269Z",
  "fullName": "user abc",
  "gender": "female",
  "govtId": "12345-1234567-2",
  "govtIdType": "nicop",
  "guardianName": "user abc husband",
  "id": Any<Number>,
  "meta": Object {},
  "nationality": "Pakistani",
  "passportNumber": "12345-1234567-1",
  "relationType": "husband",
  "updatedAt": Any<String>,
}
`;

exports[`customers Should List all Customers 3`] = `
Object {
  "createdAt": Any<String>,
  "dateOfBirth": "2021-02-19T05:20:01.269Z",
  "fullName": "user xyz",
  "gender": "male",
  "govtId": "12345-1234567-3",
  "govtIdType": "poc",
  "guardianName": "user abc father",
  "id": Any<Number>,
  "meta": Object {},
  "nationality": "Pakistani",
  "passportNumber": "12345-7654321-1",
  "relationType": "father",
  "updatedAt": Any<String>,
}
`;

exports[`customers Should List all Customers 4`] = `
Object {
  "createdAt": Any<String>,
  "dateOfBirth": "2021-02-19T05:20:01.269Z",
  "fullName": "user abc",
  "gender": "female",
  "govtId": "12345-1234567-4",
  "govtIdType": "poc",
  "guardianName": "user abc husband",
  "id": Any<Number>,
  "meta": Object {},
  "nationality": "Pakistani",
  "passportNumber": "12345-1234567-1",
  "relationType": "husband",
  "updatedAt": Any<String>,
}
`;

exports[`customers Should create a new Customer 1`] = `
Object {
  "createdAt": Any<String>,
  "dateOfBirth": "2021-02-19T05:20:01.269Z",
  "fullName": "Test Customer",
  "gender": "male",
  "govtId": "88888-8888888-8",
  "govtIdType": "cnic",
  "guardianName": "Test Customer Guardian",
  "id": Any<Number>,
  "meta": null,
  "nationality": "Pakistani",
  "passportNumber": "11111-1111111-1",
  "relationType": "father",
  "updatedAt": Any<String>,
}
`;
